src/Orleans/Core/ConcurrentHashSet.cs

```diff
@@ -0,0 +1,149 @@
+﻿/*
+Project Orleans Cloud Service SDK ver. 1.0
+ 
+Copyright (c) Microsoft Corporation
+ 
+All rights reserved.
+ 
+MIT License
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
+associated documentation files (the ""Software""), to deal in the Software without restriction,
+including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
+subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.Serialization;
+using System.Threading;
+
+namespace Orleans
+{
+
+    using Orleans.Extensions;
+
+    public sealed class ConcurrentHashSet<T> : 
+        ISerializable, 
+        IDeserializationCallback, 
+        ISet<T>
+    {
+        private readonly HashSet<T> hashSet = new HashSet<T>();
+        private readonly ReaderWriterLockSlim rwl = new ReaderWriterLockSlim();
+         
+        public ReaderWriterLockSlim Lock {  get {  return rwl; } }
+        
+        public IEnumerator<T> GetEnumerator()
+        {
+            return hashSet.GetEnumerator();
```

Comment:
For thread safety, don't we need to read lock copying the contents, then enumerate the copy?